### PR TITLE
Components in managers

### LINKED ISSUE #163: Move component ownership to managers
Store components in their respective manager rather than storing everything in World.

Also, store components continuously in memory rather than dynamically allocating them all over the place.

### PR DESCRIPTION
**Pull request by Sajdkick**
fixes #163

### CODE REVIEW

File: src/Engine/Manager/Managers.cpp [hunk 1]
```diff
@@ -39,4 +46,92 @@ void Hub::ShutDown() {
...
+        renderManager->AddComponent(world, component, componentType);
+    else if (name == "class Component::Script *")
+        scriptManager->AddComponent(world, component, componentType);
+    else if (name == "class Component::SoundSource *")
+        soundManager->AddComponent(world, component, componentType);
+    else if (name == "class Component::SpotLight *")
+        renderManager->AddComponent(world, component, componentType);
+    else
+        Log() << name << " not assigned to a manager!" << "\n";
+
```
Comment (Sajdkick) [inline src/Engine/Manager/Managers.cpp]: This is a bad solution, but if we want to rework the entire system it's good enough as a temporary thing.

File: src/Engine/Manager/SuperManager.hpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Component {
+    class SuperComponent;
+}
+
+class World;
+
+/// %Component which all components inherit.
+class SuperManager {
+public:
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SuperManager.hpp]: Scope should be indented as per code standard.

File: src/Engine/Manager/SuperManager.hpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#include <map>
+#include <typeinfo>
+
+namespace Component {
+    class SuperComponent;
+}
+
+class World;
+
+/// %Component which all components inherit.
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SuperManager.hpp]: This is not a component, though.

File: src/Engine/Manager/SuperManager.hpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+class SuperManager {
+public:
+
+    /// Destructor.
+    virtual ~SuperManager();
+
+    /// Add a component to the managers component list.
+    /**
+     * @param component The component to add.
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SuperManager.hpp]: world parameter undocumented

File: src/Engine/Manager/SuperManager.hpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+     * @param componentType The type of the component to add.
+     */
+    void AddComponent(World* world, Component::SuperComponent* component, const std::type_info* componentType);
+
+    /// Gets a list of the requested component type.
+    /**
+     * @return A list of pointers to all components of the world.
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SuperManager.hpp]: world parameter undocumented

File: src/Engine/Manager/SuperManager.hpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    template<typename T> std::vector<T*>& GetComponents(World* world);
+
+    /// Clears the components of a certain world.
+    void ClearComponents(World* world);
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SuperManager.hpp]: world parameter undocumented

File: src/Engine/Manager/SuperManager.hpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// Clears the entire component list.
+    void ClearAllComponents();
+
+    /// Clears the killed components from the list.
+    void ClearKilledComponents(World* world);
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SuperManager.hpp]: world parameter undocumented

File: src/Engine/Manager/SuperManager.hpp [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+protected:
+    // Map containing list of components.
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SuperManager.hpp]: Protected members should have doxygen comments.

File: src/Engine/Manager/Managers.cpp [hunk 9]
```diff
@@ -39,4 +46,92 @@ void Hub::ShutDown() {
...
+    shutdown = true;
+
+}
+
+void Hub::AddComponent(World* world, Component::SuperComponent* component, const std::type_info* componentType) {
+
+    std::string name = componentType->name();
+
+    if (name == "class Component::Animation *")
```
Comment (Chainsawkitten) [inline src/Engine/Manager/Managers.cpp]: These strings are compiler (and compiler version) specific. This will only work on MSVC15 specifically.

**Pull request merged**